### PR TITLE
ASoC: SOF: ipc4: specify BE format for SDW codecs

### PR DESCRIPTION
**Pull request by bardliao**
Some codecs don't support 32 bit format. We can specify the format in topology in that case.

### CODE REVIEW

File: sound/soc/sof/ipc4-topology.c [hunk 1]
```diff
@@ -502,6 +504,8 @@ static int sof_ipc4_widget_setup_comp_dai(struct snd_sof_widget *swidget)
 		goto free_available_fmt;
 	}
 
+	/* Set initial value, so that we won't fixup the format if the format is not specified */
+	ipc4_copier->frame_fmt = -1;
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c:508]: Not following @bardliao. 

The codec knows nothing about the Intel copier. the only thing that matters is the data port word length, that's a different configuration.

It's similar to I2S, you can have 32 bit words provided to the IP but only 24 provided on the link.

Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:508]: BTW frame_fmt is a u32

File: sound/soc/sof/ipc4-pcm.c [hunk 2]
```diff
@@ -560,6 +561,23 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
...
+			snd_mask_none(fmt);
+			snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S24_LE);
+			break;
+		case SOF_IPC_FRAME_S32_LE:
+			snd_mask_none(fmt);
+			snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
+			break;
+		default:
+			break;
+		}
```
Comment (ranj063) [inline sound/soc/sof/ipc4-pcm.c]: maybe you can condense  it to 
```
switch (ipc4_copier->frame_fmt) {
	case SOF_IPC_FRAME_S16_LE:
	case SOF_IPC_FRAME_S24_4LE:
	case SOF_IPC_FRAME_S32_LE:
		snd_mask_none(fmt);
		snd_mask_set_format(fmt, ipc4_copier->frame_fmt);
		break;
	default:
		break;
	}
```

Comment (bardliao) [reply sound/soc/sof/ipc4-pcm.c]: > maybe you can condense it to
> 
> ```
> switch (ipc4_copier->frame_fmt) {
> 	case SOF_IPC_FRAME_S16_LE:
> 	case SOF_IPC_FRAME_S24_4LE:
> 	case SOF_IPC_FRAME_S32_LE:
> 		snd_mask_none(fmt);
> 		snd_mask_set_format(fmt, ipc4_copier->frame_fmt);
> 		break;
> 	default:
> 		break;
> 	}
> ```

No, SOF_IPC_FRAME_S16_LE and SNDRV_PCM_FORMAT_S16_LE are not identical

File: sound/soc/sof/ipc4-topology.c (hunk 1 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c:508]: > BTW frame_fmt is a u32

Yes, but I need an invalid value.

Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c:508]: > Not following @bardliao.
> 
> The codec knows nothing about the Intel copier. the only thing that matters is the data port word length, that's a different configuration.

copier->frame_fmt is used to specify the BE format only. We don't pass frame_fmt to firmware. 

> 
> It's similar to I2S, you can have 32 bit words provided to the IP but only 24 provided on the link.

Right, and I am doing the same thing here. frame_fmt is the format on the link and `struct sof_ipc4_audio_format` is used in the IP. To be clear, I am trying to change the value of hw_params.

Comment (plbossart) [reply sound/soc/sof/ipc4-topology.c:508]: > To be clear, I am trying to change the value of hw_params.

Humm, sorry @bardliao you lost me.

Can we go back to the problem statement?

````
Some codecs don't support 32 bit format.
````

Does this mean the codec cannot support the 32-bit Word Length?

Or does this mean the codec does not support S32_LE with 24 valid bits?

Different levels, not sure if this is a hw_params issue or a SoundWire/link issue.

Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c:508]: > > To be clear, I am trying to change the value of hw_params.
> 
> Humm, sorry @bardliao you lost me.
> 
> Can we go back to the problem statement?
> 
> ```
> Some codecs don't support 32 bit format.
> ```
> 
> Does this mean the codec cannot support the 32-bit Word Length?
> 
> Or does this mean the codec does not support S32_LE with 24 valid bits?
> 
> Different levels, not sure if this is a hw_params issue or a SoundWire/link issue.

The issue is that the codec cannot support 32-bit word length.

File: sound/soc/sof/ipc4-pcm.c (hunk 2 above)
Comment (btian1) [reply sound/soc/sof/ipc4-pcm.c]: at least can use a var to store the format, then set once instead of function calling in each case.

Comment (bardliao) [reply sound/soc/sof/ipc4-pcm.c]: > at least can use a var to store the format, then set once instead of function calling in each case.

The issue is that snd_mask_none(fmt); can only be call when snd_mask_set_format() is called. So, it will look like:
```
		switch (ipc4_copier->frame_fmt) {
		case SOF_IPC_FRAME_S16_LE:
			var = SNDRV_PCM_FORMAT_S16_LE);
			break;
		case SOF_IPC_FRAME_S24_4LE:
			var = SNDRV_PCM_FORMAT_S24_LE);
			break;
		case SOF_IPC_FRAME_S32_LE:
			var = SNDRV_PCM_FORMAT_S32_LE);
			break;
		default:
			break;
		}
		if (var) {
			snd_mask_none(fmt);
			snd_mask_set_format(fmt, var);
		}
```
IMHO, I don't see any benefit on it. We add a variable and a test, but the code doesn't become simpler.

File: sound/soc/sof/ipc4-topology.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-topology.c:508]: Let me try again @bardliao (adding @ranj063)

If we had an I2S codec and wanted to render in 16 bits only, how would we specify this property for the BE?
a) tdm_slot_width
b) input/output format
c) something else?

my point is that I don't see a need to have an SoundWire-specific property for the BE, we have a similar problem with I2S and we should try to use common encoding/patterns to describe the same things. So far we didn't need this frame_fmt in the copier, so why would we need it now?

Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c:508]: @plbossart snd_mask_set_format() works for I2S, too. It is totally fine to move the snd_mask_set_format section out of switch (ipc4_copier->dai_type) {}. We did it for IPC3. See [switch (private->comp_dai->config.frame_fmt)](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc3-pcm.c#L237) and I use the same topology token as IPC3 did [SOF_TKN_COMP_FORMAT](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc3-topology.c#L163)

Comment (plbossart) [reply sound/soc/sof/ipc4-topology.c:508]: Not following your last comment @bardliao, what is your suggestion?

Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:508]: @plbossart are you suggesting that we use the bit depth and valid bit depth in then output audio format that are already specified in topology to decide whether to perform fix up here instead? That would need a bit more code to derive the frame format based on bit depths but would need to additional token in topology

Comment (plbossart) [reply sound/soc/sof/ipc4-topology.c:508]: @ranj063 @bardliao My worry is that we already have 2 places where the format restrictions are specified:

a) in the codec driver, e.g. for max98363.c:
#define MAX98363_FORMATS (SNDRV_PCM_FMTBIT_S16_LE | SNDRV_PCM_FMTBIT_S24_LE)

b) in the copier 'valid_bit_depth' copier format

the topology SOF_TKN_COMP_FORMAT would be a 3rd case which needs to be consistent with the first two.
So all I am asking is: can we avoid this new definition?

I am ok if there are technical arguments and the complexity is not worth it, but when we have such cases we need to balance the ease of using a single piece of information v. the risk of screwing-up with two misaligned values. We've had this problem with SSP in the past where we had the DAI FIFO format and bits on the wire that were inconsistent.

Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c:508]: @plbossart @ranj063 Alternative PR is https://github.com/thesofproject/linux/pull/4446 which set valid_bit_depth as fixup if the DAI supports only one input/output format depends on the direction.

> @ranj063 @bardliao My worry is that we already have 2 places where the format restrictions are specified:

> a) in the codec driver, e.g. for max98363.c: #define MAX98363_FORMATS (SNDRV_PCM_FMTBIT_S16_LE | SNDRV_PCM_FMTBIT_S24_LE)

I am not sure who checks this. Currectly, ASoC will still use S32 format if we play a 32 bit format audio file.
I expect that ASoC will check the available formats in CPU and CODEC DAIs and select a format that both CPU and CODEC DAIs can support. But, I was wrong. ASoC didn't do it.

> 
> b) in the copier 'valid_bit_depth' copier format

Isn't that just inside SOF format? I meant SOF can work with the valid_bit_depth format but use a different format to transfer physically.

> 
> the topology SOF_TKN_COMP_FORMAT would be a 3rd case which needs to be consistent with the first two. So all I am asking is: can we avoid this new definition?

Will https://github.com/thesofproject/linux/pull/4446 be better? No new topology token is used, just uses valid_bit_depth.

> 
> I am ok if there are technical arguments and the complexity is not worth it, but when we have such cases we need to balance the ease of using a single piece of information v. the risk of screwing-up with two misaligned values. We've had this problem with SSP in the past where we had the DAI FIFO format and bits on the wire that were inconsistent.

**Comment by bardliao** [comment]
Replaced by #4446
